include/boost/process/v2/impl/pid.ipp

```diff
@@ -22,6 +65,603 @@ pid_type current_pid() {return ::GetCurrentProcessId();}
+            }
+        }
+    }
+    else
+        ec = detail::get_last_error();
+    kvm_close(kd);
+    return vec;
+}
+
+#elif defined(__OpenBSD__)
+
+std::vector<pid_type> all_pids(std::error_code & ec)
+{
+    std::vector<pid_type> vec;
+    int cntp = 0;
+    kinfo_proc *proc_info = nullptr;
+    kd = kvm_openfiles(nullptr, nullptr, nullptr, KVM_NO_FILES, nullptr);
+    if (!kd)
+    {
+        ec = detail::get_last_error();
+        return vec;
+    } 
+    if ((proc_info = kvm_getprocs(kd, KERN_PROC_ALL, 0, sizeof(struct kinfo_proc), &cntp)))
+    {
+        vec.reserve(vec.size() + cntp);
+        for (int i = cntp - 1; i >= 0; i--) {
+        {
+            if (proc_info[i].kp_pid >= 0) 
+            {
+                vec.push_back(proc_info[i].kp_pid);
+            }
+        }
+    }
+    else
+        ec = detail::get_last_error();
+    kvm_close(kd);
+    return vec;
+}
+
+std::vector<pid_type> parent_pid(pid_type pid, std::error_code & ec) {
+    std::vector<pid_type> vec;
+    int cntp = 0;
+    kinfo_proc *proc_info = nullptr;
+    kd = kvm_openfiles(nullptr, nullptr, nullptr, KVM_NO_FILES, nullptr);
+    if (!kd)
+    {
+        ec = detail::get_last_error();
+        return vec;
+    }
+    if ((proc_info = kvm_getprocs(kd, KERN_PROC_PID, pid, sizeof(struct kinfo_proc), &cntp)))
+    {
+        vec.push_back(proc_info->p_ppid);
+    }
+    else
+        ec = detail::get_last_error();
+    kvm_close(kd);
+    return vec;
+}
+
+std::vector<pid_type> child_pids(pid_type pid, std::error_code & ec) {
+    std::vector<pid_type> vec;
+    int cntp = 0;
+    kinfo_proc *proc_info = nullptr;
+    kd = kvm_openfiles(nullptr, nullptr, nullptr, KVM_NO_FILES, nullptr); 
+    if (!kd) 
+    {
+        ec = detail::get_last_error();
+        return vec;
+    }
+    if ((proc_info = kvm_getprocs(kd, KERN_PROC_ALL, 0, sizeof(struct kinfo_proc), &cntp)))
+    {
+        for (int i = cntp - 1; i >= 0; i--)
+        {
+            if (proc_info[i].p_ppid == pid)
+            {
+                vec.push_back(proc_info[i].p_pid);
+            }
+        }
+    }
+    else
+        ec = detail::get_last_error();
+    kvm_close(kd);
+    return vec;
+}
+
+
+#elif defined(__sun)
+
+std::vector<pid_type> all_pids(std::error_code & ec)
+{
+    std::vector<pid_type> vec;
+    struct pid cur_pid;
+    proc *proc_info = nullptr;
+    kd = kvm_open(nullptr, nullptr, nullptr, O_RDONLY, nullptr);
+    if (!kd)
+    {
+        ec = detail::get_last_error();
+        return vec;
+    } 
+    while ((proc_info = kvm_nextproc(kd)))
+    {
+        if (kvm_kread(kd, (std::uintptr_t)proc_info->p_pidp, &cur_pid, sizeof(cur_pid)) != -1)
+        {
+            vec.insert(vec.begin(), cur_pid.pid_id);
+        }
+        else
+        {
+            ec = detail::get_last_error();
+            break;
+        }
+    }
+    kvm_close(kd);
+    return vec;
+}
+
+std::vector<pid_type> parent_pid(pid_type pid, std::error_code & ec) {
```

Comment:
why is this returning a vector and not just one `pid_type` ?

Reply:
in the case of error, it shouldn't return a pid at all. If you have a suggestion on an alternate way of handling that, I'm all ears.

Reply:
I think `-1` is fair - the value is nonsense if `ec` got set.

Reply:
-1 is not possible on windows because it is unsigned long. although we could do static_cast<pid_type>(-1) which i guess is basically what you meant